plugins/synaptics-cape/fu-synaptics-cape-firmware.c

```diff
@@ -0,0 +1,215 @@
+}
+
+static gboolean
+fu_synaptics_cape_firmware_parse_header(FuSynapticsCapeFirmware *self,
+					FuFirmware *firmware,
+					GBytes *fw,
+					GError **error)
+{
+	gsize bufsz = 0x0;
+	const guint8 *buf = g_bytes_get_data(fw, &bufsz);
+	g_autoptr(GBytes) fw_hdr = NULL;
+	g_autoptr(FuFirmware) img_hdr = fu_firmware_new();
+
+	/*the input fw image size shoud be the same as header size */
+	if (bufsz < sizeof(FuCapeHidFileHeader)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "not enough data to parse header");
+		return FALSE;
+	}
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VID,
+					&self->vid,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_PID,
+					&self->pid,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VER_X,
+					&self->version_x,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VER_Y,
+					&self->version_y,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VER_Z,
+					&self->version_z,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	self->version = g_strdup_printf("%u.%u.%u.%u",
+					self->version_z,
+					self->version_y,
+					self->version_x,
+					self->version_w);
+	fu_firmware_set_version(FU_FIRMWARE(self), self->version);
+
+	fw_hdr = fu_common_bytes_new_offset(fw, 0, sizeof(FuCapeHidFwCmdUpdateStartPar), error);
+	if (fw_hdr == NULL)
+		return FALSE;
+
+	fu_firmware_set_id(img_hdr, FU_FIRMWARE_ID_HEADER);
+	fu_firmware_set_bytes(img_hdr, fw_hdr);
+	fu_firmware_add_image(firmware, img_hdr);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_cape_firmware_parse(FuFirmware *firmware,
+				 GBytes *fw,
+				 guint64 addr_start,
+				 guint64 addr_end,
+				 FwupdInstallFlags flags,
+				 GError **error)
+{
+	FuSynapticsCapeFirmware *self = FU_SYNAPTICS_CAPE_FIRMWARE(firmware);
+	const gsize bufsz = g_bytes_get_size(fw);
+	g_autoptr(GBytes) fw_header = NULL;
+	g_autoptr(GBytes) fw_body = NULL;
+	const gsize headsz = sizeof(FuCapeHidFileHeader);
+
+	/* check minimum size */
+	if (bufsz < sizeof(FuCapeHidFileHeader)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "not enough data to parse header, size ");
+		return FALSE;
+	}
+
+	if ((guint32)bufsz % 4 != 0) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "data not aligned to 32 bits");
+		return FALSE;
+	}
+
+	fw_header = g_bytes_new_from_bytes(fw, 0x0, headsz);
+
+	if (fw_header == NULL) {
+		g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE, "out of memory");
```

Comment:
Can this actually happen?  `g_bytes_new_from_bytes` is supposed to be a subsection of the existing `GBytes`

Reply:
agree; it can't really happen as it's really just another ref on the existing data but with a different "window"

Reply:
Thanks for catching this.